### PR TITLE
Equip FEEvaluation with code path without template on polynomial degree

### PR DESCRIPTION
**Pull request by kronbichler**
This commit equips `FEEvaluation` with the ability to use the template parameter `-1` as an indication that the polynomial degree should be read from the underlying `FiniteElement` object that was sent into `MatrixFree`. This has been realized by creating a specialized evaluator object which takes the run-time values of `fe_degree` and `n_q_points_1d` rather than the given compile-time values. Since this is through a template specialization class, this comes without overhead. Of course, the version without templates is slower: step-37 with 2.1m DoFs solves the linear system in 2.8s on a single core with templates on `FEEvaluation`, but it takes 7s without templates.

In order to provide FEEvaluation with a flexible amount of scratch data, I also introduced a `scratch_data` memory field in `MatrixFree` that gets re-used between different calls (and is thread safe). This scratch data is used in all invocations of `FEEvaluation` and also allows to run with very large polynomial degrees. Previously, one would run into a stack overflow at polynomial degree 16 or so in 3D. I tested this with polynomial degree 20 in a test... ;-)

Finally, the availability of a `default` code path in `FEEvaluation` means that we can treat the *unusual* polynomial degrees in `VectorTools::project` by the new code.

### CODE REVIEW

**Comment by kronbichler** [comment]
Avoid too much work in compiling `vector_tools_xxx` is a nice side-effect of this PR. Since this is a pretty safe change code-wise in my opinion, this should be fine for inclusion in the 8.5 release.

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 1]
```diff
@@ -1537,6 +1582,37 @@ class FEEvaluationAccess<1,1,Number> : public FEEvaluationBase<1,1,Number>
...
+ *
+ * A flexible choice of the polynomial degree based on the information in the
+ * element passed to the initialization is also supported by FEEvaluation,
+ * even though it runs two to three times slower. For this, set the template
+ * parameter of the polynomial degree to -1 (and choose an arbitrary number
+ * for the number of quadrature points), which switches to the other code
+ * path. Internally, an evaluator object with template specialization for -1
+ * is invoked that runs according to run-time bounds, whereas the default case
+ * with compile-time bounds given through fe_degree selects another template
+ * class without comprising efficiency.
```
Comment (davydden) [inline include/deal.II/matrix_free/fe_evaluation.h]: compromising?

Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h]: Changed.

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 2]
```diff
@@ -1649,6 +1725,37 @@ class FEEvaluationAccess<1,1,Number> : public FEEvaluationBase<1,1,Number>
  * space. Currently, other finite elements cannot be treated with the
  * matrix-free concept.
  *
+ * <h4>Degree of finite element as a compile-time parameter</h4>
+ *
+ * The default usage model of FEEvaluation expects the polynomial degree to be
+ * given as a template parameter. This requirement guarantees maximum
+ * efficiency: The sum factorization evaluation performs a number of short 1D
+ * loops of length equal to the polynomial degree. If the loop bounds are
```
Comment (bangerth) [inline include/deal.II/matrix_free/fe_evaluation.h]: same as above

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 3]
```diff
@@ -1811,7 +1918,9 @@ class FEEvaluationAccess<1,1,Number> : public FEEvaluationBase<1,1,Number>
  * @tparam dim Dimension in which this class is to be used
  *
  * @tparam fe_degree Degree of the tensor product finite element with
- * fe_degree+1 degrees of freedom per coordinate direction
+ * fe_degree+1 degrees of freedom per coordinate direction. Can be set to -1
+ * if the degree is not known at compile time, but performance will usually be
+ * by a factor of 2-3 lower.
```
Comment (bangerth) [inline include/deal.II/matrix_free/fe_evaluation.h]: will usually be WORSE by a factor

**Comment by bangerth** [comment]
Have you though about possibly not using `-1` as the "special" value, but `numbers::invalid_unsigned_int`? Or did you choose `-1` because the current type of that template argument is `int` instead of `unsigned int`?

**Comment by bangerth** [comment]
Also, would it be hard to simply clone *all* matrix-free tests and replace whatever template argument they use to `-1`? They should all continue to produce the same output, so that would be relatively little effort...

**Comment by davydden** [comment]
one can also use `0`, but probably `-1` is more telling.

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 4]
```diff
@@ -582,6 +587,16 @@ class FEEvaluationBase
...
 
+  /**
+   * Returns a pointer to internal memory for temporary use. Note that some of
+   * this memory is overwritten during evaluate() and integrate() calls so do
+   * not assume it to be stable over those calls. The maximum size you can
+   * write into is (n_components+2)*dofs_per_cell+2*n_q_points, access beyond
+   * is past-end-of-array.
+   */
+  VectorizedArray<Number> *
+  begin_scratch_data() const;
```
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h]: That's a good idea. I will change that.

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 5]
```diff
@@ -678,24 +693,41 @@ class FEEvaluationBase
...
-   * Internal data fields that store the values. Derived classes will know the
-   * length of all arrays at compile time and allocate the memory on the
-   * stack. This makes it possible to cheaply set up a FEEvaluation object and
-   * write thread-safe programs by letting each thread own a private object of
-   * this type. In this base class, only pointers to the actual data are
-   * stored.
-   *
+   * This is the general array for all data fields.
+   */
+  AlignedVector<VectorizedArray<Number> > *scratch_data_array;
```
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h]: Yes, it indeed matches with the other types. The other reason for using a plain pointer here is that this data field usually comes from a thread local storage pool, so to hand it back and mark it as available again, I do not see a better solution than the plain pointer. This is also the way we use the list of objects in `WorkStream`. I will add a note that this usage is similar to how `WorkStream` handles the scratch data.

**Comment by kronbichler** [comment]
> Have you thought about possibly not using `-1` as the "special" value, but `numbers::invalid_unsigned_int`? Or did you choose -1 because the current type of that template argument is `int` instead of `unsigned int`?

Indeed, the template parameter is an integer so `-1` appears to be the natural choice to me. And `numbers::invalid_unsigned_int` is really just `-1` cast to unsigned int. I could also change it but then I would have to write `unsigned int` for the template parameter... (It would have saved me a handful `static_cast` operations to silence compiler warnings of signed versus unsigned comparisons, though.)

> Also, would it be hard to simply clone all matrix-free tests and replace whatever template argument they use to `-1`? They should all continue to produce the same output, so that would be relatively little effort...

This is what I did for three of the tests I added. Since the only difference between the templated and `-1` case is a different kernel in the inside `internal::EvaluatorTensorProduct<evaluate_general,dim,-1,0,Number>` as well as some data fields that are initialized from the scratch memory, I think that the test cases cover all functionality. I am a bit reluctant to clone all tests because a number of tests are quite long-running (for example because they compute a sparse matrix for a reference solution) and increase the overall run time of the test suite by approximately 5 minutes on 24 cores.

**Comment by kronbichler** [comment]
> one can also use `0`, but probably `-1` is more telling.

Actually, a polynomial degree of 0 is a valid polynomial degree, think of `FE_DGQ(0)`. I have done runs with those elements and even solved a Laplacian - which only consists of the DG-SIP penalty terms in that case, obviously. But the evaluation approach is clearly not efficient for that case.

**Comment by davydden** [comment]
> Actually, a polynomial degree of 0 is a valid polynomial degree, think of FE_DGQ(0).

right, i completely forgot about those FEs.

**Comment by kronbichler** [comment]
I now also added the non-templated version to avoid `Assert(false, ExcNotImplemented())` cases in `MGTransferMatrixFree`.

File: include/deal.II/matrix_free/fe_evaluation.h (hunk 5 above)
Comment (bangerth) [inline include/deal.II/matrix_free/fe_evaluation.h]: OK, thanks for the explanation. I guess I'm neither fond of bare `new/delete` nor of pointer members pointing to either data that has been statically allocated *or* dynamically allocated. Both just lends itself to memory leaks or memory corruption (if you accidentally delete something you don't own).

But that's all for a different time to clean up. No worries.

**Comment by bangerth** [comment]
OK, thanks for tackling this! I have no further comments or requests.

I do have a question, though: You cite a factor of 2-3 in performance. Is this for low- or high-order elements? Does the difference disappear for sufficiently high order elements?

**Comment by kronbichler** [comment]
> You cite a factor of 2-3 in performance. Is this for low- or high-order elements? Does the difference disappear for sufficiently high order elements?

Let's look into this. I'll post the results in an hour.

**Comment by bangerth** [comment]
Thanks.

Here's why I'm curious: Let's assume that the difference goes away at p=5. Then we could contemplate something along these lines:
```
  template <int degree>
  void the_function_that_actually_does_the_work (...) { ... }    // currently publicly available and
                                                                                                // instantiated by the user; we may
                                                                                                // now hide it because...

  template <int degree>
  void the_function_that_provides_the_public_interface (...)   // ...we now have this one!
  {
     the_function_that_actually_does_the_work< (degree >= 5 ? -1 : degree) > (...)
  }   
```
This would limit instantiation of the functions that actually do the real work to degree=0..5, and beyond that we would just use the generic implementation. It may help us with the compile time issues.

I'm not proposing that you implement such a scheme. I'm just curious whether that would be an option :-)

**Comment by kronbichler** [comment]
It turns out that Katharina and I did a good job when implementing these kernels: The difference does not go away, not even for polynomial degree equal to 30. Here are results from the 3D Laplacian:

times_over_degree_matvec

The gap starts out at approximately 2.4 at p=1 and reaches a maximum at p=2 with 3.0. The non-templated version is then between 2.5x (for p=9) and 2.9 times slower until degree 17 but the goes up to 3.2 for degree 30. Note that the templated version can utilize a trick that cuts the operation count into slightly more than half (called even-odd decomposition; it relies on the fact that shape functions and quadrature points are symmetric about the center of the element) that is not available for the new code. While I could have implemented that as well, it would not really pay off since there is still a factor 1.6-1.9 left.

**Comment by davydden** [comment]
@kronbichler maybe add this picture to class documentation somewhere?

**Comment by kronbichler** [comment]
Great idea @davydden. I will add one more commit and upload the picture to the homepage.

**Comment by bangerth** [comment]
On 02/17/2017 11:58 AM, Martin Kronbichler wrote:
> It turns out that Katharina and I did a good job when implementing these
> kernels

Everyone who knows the two of you had no doubt about this anyway :-)

Thanks for trying this out, that's interesting data!

**Comment by bangerth** [comment]
On 02/17/2017 12:41 PM, Martin Kronbichler wrote:
> Great idea @davydden <https://github.com/davydden>. I will add one more
> commit and upload the picture to the homepage.

I bet the picture is small enough that it can be included in 
doc/doxygen/images (or whatever the correct directory is) just fine.

**Comment by kronbichler** [comment]
That's what I did, I included the picture in the doxygen folder.

**Comment by kronbichler** [comment]
I changed the axis in the picture slightly to have a finer mesh on the y axis. This should now be ready for merging.

**Pull request merged**